### PR TITLE
[SPARK-49791][SQL] Make DelegatingCatalogExtension more extendable

### PR DESCRIPTION
**Pull request by cloud-fan**
### What changes were proposed in this pull request?

This PR updates `DelegatingCatalogExtension` so that it's more extendable
- `initialize` becomes not final, so that sub-classes can overwrite it
- `delegate` becomes `protected`, so that sub-classes can access it

In addition, this PR fixes a mistake that `DelegatingCatalogExtension` is just a convenient default implementation, it's actually the `CatalogExtension` interface that indicates this catalog implementation will delegate requests to the Spark session catalog. https://github.com/apache/spark/pull/47724 should use `CatalogExtension` instead. 

### Why are the changes needed?

Unblock the Iceberg extension.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

existing tests

### Was this patch authored or co-authored using generative AI tooling?

no

### CODE REVIEW

**Review by yaooqinn — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/DelegatingCatalogExtension.java [hunk 1]
```diff
@@ -51,7 +51,9 @@ public String name() {
   }
 
   @Override
-  public final void initialize(String name, CaseInsensitiveStringMap options) {}
+  public void initialize(String name, CaseInsensitiveStringMap options) {
+    delegate.initialize(name, options);
+  }
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/DelegatingCatalogExtension.java:56]:
```suggestion
  public void initialize(String name, CaseInsensitiveStringMap options) {}
```

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

To @cloud-fan , 
- Do we need to `@DeveloperApi` annotation additionally?
- Do you want to backport this to `branch-3.5`?

**Review by RussellSpitzer — APPROVED** [review]

**Comment by cloud-fan** [comment]
Hi @dongjoon-hyun , the entire `DelegatingCatalogExtension` class is marked as Evolving, which already means semi-public APIs.

**Comment by dongjoon-hyun** [comment]
Merged to master/3.5! Thank you, @cloud-fan and all!

**Comment by zhouyuan** [comment]
Hi @dongjoon-hyun @cloud-fan Could you please let me know if there is a plan or scheduled release date for Spark-354?

thanks, -yuan

**Pull request closed without merging**